main.py

```diff
@@ -0,0 +1,356 @@
+        index = photo_name.index('.jpg')
+        thumbnail_key = '{}{}{}'.format(photo_name[:index],
+                                        generation_number,
+                                        photo_name[index:])
+
+        # Create the Notification using the received information.
+        new_notification = create_notification(
+            photo_name,
+            event_type,
+            generation_number,
+            overwrote_generation=overwrote_generation,
+            overwritten_by_generation=overwritten_by_generation)
+
+        # If the new_notification already has been stored, it is a
+        # repeat and can be ignored.
+        exists_notification = Notification.query(
+            Notification.message == new_notification.message,
+            Notification.generation == new_notification.generation).get()
+
+        if exists_notification:
+            return
+
+        # Don't act for metadata update events.
+        if new_notification.message is None:
+            return
+
+        # Store new_notification in datastore.
+        new_notification.put()
+
+        # For create events: shrink the photo to thumbnail size,
+        # store the thumbnail in GCS, and create a ThumbnailReference
+        # and store it in Datastore.
+        if event_type == 'OBJECT_FINALIZE':
+            thumbnail = create_thumbnail(photo_name)
+            store_thumbnail_in_gcs(thumbnail_key, thumbnail)
+            original_photo = get_original_url(photo_name, generation_number)
+            uri = '{}{}{}{}'.format('gs://',
+                                    PHOTO_BUCKET,
+                                    '/',
+                                    photo_name)
+            labels = get_labels(uri, photo_name)
+            thumbnail_reference = ThumbnailReference(
+                thumbnail_name=photo_name,
+                thumbnail_key=thumbnail_key,
+                labels=list(labels),
+                original_photo=original_photo)
+            thumbnail_reference.put()
+
+        # For delete/archive events: delete the thumbnail from GCS
+        # and delete the ThumbnailReference.
+        elif event_type == 'OBJECT_DELETE' or event_type == 'OBJECT_ARCHIVE':
+            delete_thumbnail(thumbnail_key)
+
+
+# Create a Notification to be stored in Datastore.
+def create_notification(photo_name,
+                        event_type,
+                        generation,
+                        overwrote_generation=None,
+                        overwritten_by_generation=None):
+    if event_type == 'OBJECT_FINALIZE':
+        if overwrote_generation is not None:
+            message = '{} {}'.format(
+                photo_name,
+                'was uploaded and overwrote an older version of itself.')
+        else:
+            message = '{} {}'.format(photo_name, 'was uploaded.')
+    elif event_type == 'OBJECT_ARCHIVE':
+        if overwritten_by_generation is not None:
+            message = '{} {}'.format(photo_name,
```

Comment:
A nit about indention continuation style: I *strongly* prefer to avoid "hanging indents" like this:

```
message = '{} {}'.format(photo_name,
                                         'was overwritten by a newer version')
```

And instead prefer to line break at the earliest possible place and just indent once:

```
message = '{} {}'.format(
    photo_name, 'was overwritten by a newer version')
```

or

```
message = '{} {}'.format(
    photo_name,
    'was overwritten by a newer version')
```